crates/bevy_ui/src/keyboard_navigation.rs

```diff
@@ -0,0 +1,173 @@
+        let first_index = current_index.map(|index| index + 1).unwrap_or_default();
+
+        let wrapping_nodes_iterator = ui_nodes
+            .iter()
+            .skip(first_index)
+            .chain(ui_nodes.iter().take(first_index));
+
+        wrapping_nodes_iterator.filter(can_focus).next().copied()
+    };
+
+    // Reset the clicked state
+    if new_focus != focus.entity {
+        if let Some(node) = focus
+            .entity
+            .and_then(|entity| node_query.get_mut(entity).ok())
+        {
+            if let Some(mut interaction) = node.interaction {
+                if *interaction == Interaction::Clicked {
+                    *interaction = Interaction::None;
+                }
+            }
+        }
+    }
+
+    focus.set_if_neq(Focus {
+        entity: new_focus,
+        focus_visible: true,
+    });
+}
+
+/// Change the [`FocusedState`] for the specified entity
+fn set_focus_state<'a>(
+    entity: Option<Entity>,
+    focus_state: &'a mut Query<&mut FocusedState>,
+    new_state: FocusedState,
+) {
+    if let Some(mut focus_state) = entity.and_then(|entity| focus_state.get_mut(entity).ok()) {
+        focus_state.set_if_neq(new_state);
+    }
+}
+
+pub(crate) fn update_focused_state(
+    mut focus_state: Query<&mut FocusedState>,
+    focus: Res<Focus>,
+    mut old_focused_entity: Local<Option<Entity>>,
+) {
+    let new_focused_entity = focus.entity;
+
+    // Remove the interaction from the last focused entity
+    if *old_focused_entity != new_focused_entity {
+        set_focus_state(*old_focused_entity, &mut focus_state, FocusedState::None);
+    }
+
+    let focus_visible = focus.focus_visible;
+    let new_state = FocusedState::Focus { focus_visible };
+    // Set the focused interaction on the newly focused entity
+    set_focus_state(new_focused_entity, &mut focus_state, new_state);
+
+    *old_focused_entity = new_focused_entity;
+}
+
+/// Should the [`keyboard_click`] system run?
+pub(crate) fn trigger_click(keyboard_input: Res<Input<KeyCode>>) -> bool {
+    keyboard_input.just_pressed(KeyCode::Space) || keyboard_input.just_pressed(KeyCode::Return)
+}
+
+/// Trigger the [`Focused`] entity to be clicked.
+pub(crate) fn keyboard_click(mut interactions: Query<&mut Interaction>, focus: Res<Focus>) {
+    if let Some(mut interaction) = focus
+        .entity
+        .and_then(|entity| interactions.get_mut(entity).ok())
+    {
+        interaction.set_if_neq(Interaction::Clicked);
+    }
+}
+
+/// Should the [`end_keyboard_click`] system run?
+pub(crate) fn trigger_click_end(keyboard_input: Res<Input<KeyCode>>) -> bool {
+    keyboard_input.just_released(KeyCode::Space) || keyboard_input.just_released(KeyCode::Return)
+}
+
+/// Reset the clicked state.
+pub(crate) fn end_keyboard_click(mut interactions: Query<&mut Interaction>) {
+    interactions.for_each_mut(|mut interaction| {
+        if *interaction == Interaction::Clicked {
+            *interaction = Interaction::None
```

Comment:
Why make it `None` over `Hovered`? Since it's still the "focused" entity.

Reply:
`Hovered` is when the mouse is over the button. If you trigger a button click with the keyboard, then after your key press, I don't see why the button would be hovered over by the mouse?

Reply:
This makes sense. I wanted to make sure I understood the reasoning.

Reply:
I've added a comment.